src/System.IO.Ports/src/System/IO/Ports/SerialPort.Win32.cs

```diff
@@ -13,26 +16,111 @@ public static string[] GetPortNames()
+
+            // If running on Windows IoT, search and add the Serial Devices from QueryDosDevice
+            // If this is allowed to happen on Windows devices where port detection happens normally, duplicate Serial Port names are added.
+            // This causes issues with System.IO.Ports.Tests
+            // Port detection broken on Windows IoT (Does not initialise registry with COM port names), so use QueryDosDevice
+            if (RuntimeInformation.OSArchitecture == Architecture.Arm || RuntimeInformation.OSArchitecture == Architecture.Arm64)
+            {
+                // Query Interop QueryDosDevice()
+                //
+                // Is GUID class ID for COM ports https://docs.microsoft.com/en-us/windows-hardware/drivers/install/guid-devinterface-comport
+                // Typical DosDevice serial port name 'USB#VID_1B4F&PID_214F&MI_00#6&381731fd&0&0000#{86e0d1e0-8089-11d0-9ce4-08003e301f73}'
+                //
+                foreach (string dosName in QueryDosDeviceComPorts(GuidDevInterfaceComPort))
+                {
+                    resultPortNames.Add(dosName);
+                }
+            }
+            else
+            {
+                // Query registry register COM[x] ports
+                //
+                using (RegistryKey serialKey = Registry.LocalMachine.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM"))
+                {
+                    if (serialKey != null)
+                    {
+                        foreach (string valueName in serialKey.GetValueNames())
+                        {
+                            string portName = ((string)serialKey.GetValue(valueName));
+
+                            // Add the key's value (portname) if it is valid (Win10IoT places garbage in this Reg Key)
+                            // Filters corrupt chars for COM[x] devices in Registry on Win10IoT
+                            if (portName.StartsWith("COM", StringComparison.InvariantCultureIgnoreCase))
+                            {
+                                resultPortNames.Add(portName);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return resultPortNames.ToArray();
+        }
+
+        private static string[] QueryDosDeviceComPorts(string filterGuid)
+        {
+            // Build a list of all system Com Port device names.
+            // memBuff starts with a small arbitary size and dynamically expands until there is enough room for data returned from Kernal32.QueryDosDevice().
+            uint returnSize = 0;
+            int maxSize = 1024;
+            string allDevices = null;
+            const uint ERROR_INSUFFICIENT_BUFFER = 122;
 
-            using (RegistryKey serialKey = Registry.LocalMachine.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM"))
+            string[] allDevicesArray = null;
+            List<string> returnList = new List<string>();
+
+            while (returnSize == 0)
```

Comment:
The right way to do this loop is:

``` C#

while ((result = Interop.Kernel32.QueryDosDeviceW(deviceName, buffer, buffer.Length)) == 0)
{
    int error = Marshal.GetLastWin32Error();
    switch (error)
    {
        case Interop.Errors.ERROR_INSUFFICIENT_BUFFER:
            // Return and rent a larger char buffer
            break;
        default:
            throw new Win32Exception(error);
    }
}

// There are two trailing nulls, don't want an empty split
ReadOnlySpan<char> names = buffer.Slice(0, result - 2);

// Loop through and .Slice off each string, doing .Contains(guid, StringComparison.OrdinalIgnoreCase)

// Pass back matched names with `\\?\` prepended (as they're already "resolved")
```

Reply:
Struggling with this one, how to incorporate ArrayPool & unsafe pointer that needs to be re-assigned as each buffer expansion creates a new pointer reference..

Reply:
@PRIMETSS this will remind C++ manual memory management:

```csharp
byte[] array = ArrayPool.Rent(initialSize); // note size of array might be larger than requested

// too small
ArrayPool.Return(array);
array = ArrayPool.Rent(newSize);
```

+ making sure we return when something throws (i.e. try .. finally) and triple check you don't double return

Reply:
You need to rent a buffer of `char`, not `byte`. The API `[DllImport]` signature can just take `char[]` for the buffer and it will be pinned, you have nothing else to do.

Reply:
> making sure we return when something throws

If throws aren't expect to happen frequently (which is the case here) we don't need to return the buffer (and shouldn't). Adding a try block makes things slower and quickly eats away at the benefit of returning the buffer.